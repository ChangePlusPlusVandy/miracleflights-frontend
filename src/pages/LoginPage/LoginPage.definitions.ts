/* eslint-disable autofix/no-unused-vars */
export interface FormValues {
  email: string;
  password: string;
}

export enum FormValueNames {
  email = "email",
  password = "password",
}
