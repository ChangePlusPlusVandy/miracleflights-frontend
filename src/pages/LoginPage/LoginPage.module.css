/* Login.module.css */

.pageContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  height: 100vh;
}

.logo {
  align-self: flex-start; /* Align the logo to the top left */
  height: 50px;
  margin-left: 32px;
  margin-top: 24px;
}

.loginContainer {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 72px 0;
}

.loginBlock {
  background-color: var(--miracle-color-white);
  box-shadow: var(--box-shadow);
  padding: 20px;
  width: 300px;
  border-radius: 8px;
  display: flex;
  flex-direction: column;
}

.loginBlockHeader {
  font-size: 24px;
  font-weight: 400;
  margin-bottom: 8px;
  color: #333;
}

.loginBlockSubtitle {
  font-size: 14px;
  color: var(--miracle-color-medium-dark);
  margin-bottom: 16px;
}

.loginBlockContent {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.loginInputContainerUpper,
.loginInputContainerLower {
  width: 100%;
  margin-bottom: 16px;
}

.loginInput {
  padding: 8px;
  font-size: 14px;
  border: 1px solid var(--miracle-color-medium-light);
  border-radius: 4px;
  width: 100%;
  box-sizing: border-box;
  margin-bottom: 8px;
}

.loginBlockLink {
  color: var(--miracle-color-blue);
  text-decoration: none;
}

.loginBlockLink:hover {
  cursor: pointer;
}

.loginRedirect {
  margin-bottom: 16px;
  font-size: 12px;
}

.passwordToggle {
  cursor: pointer;
  display: flex;
  color: var(--miracle-color-medium-dark);
}

.Icon {
  margin: -30px 10px 0 auto;
}

.forgotPassword {
  font-size: 12px;
  color: var(--miracle-color-medium-dark);
  margin-bottom: 16px;
  margin-left: auto;
}
