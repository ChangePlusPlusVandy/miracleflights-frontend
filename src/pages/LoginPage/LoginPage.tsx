import styles from "./LoginPage.module.css";
import { FormValueNames, type FormValues } from "./LoginPage.definitions";
import Button from "../../components/Button/Button";
import { ButtonVariant } from "../../components/Button/Button.definitions";
import Icon from "../../components/CustomIcon/Icon";
import Input from "../../components/Input/Input";
import logo from "../../public/0GAGNk.tif.png";
import * as yup from "yup";
import { yupResolver } from "@hookform/resolvers/yup";
import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import { useForm } from "react-hook-form";
import { useSignIn } from "@clerk/clerk-react";

const LoginPage = () => {
  const [passwordVisible, setPasswordVisible] = useState<boolean>(false);
  const [disabled, setDisabled] = useState<boolean>(true);
  const navigate = useNavigate();
  const { signIn, setActive, isLoaded } = useSignIn();

  interface CustomError {
    code: string;
  }

  interface ErrorWithDetails {
    errors: CustomError[];
  }

  const schema = yup.object().shape({
    [FormValueNames.email]: yup
      .string()
      .email("Please enter a valid email address")
      .required("Please enter an email address"),
    [FormValueNames.password]: yup
      .string()
      .min(8, "Password must be at least 8 characters")
      .required("Please enter a password"),
  });

  const {
    register,
    handleSubmit,
    formState: { errors, isValid },
    setError: setSignInError,
  } = useForm<FormValues>({
    defaultValues: {
      [FormValueNames.email]: "",
      [FormValueNames.password]: "",
    },
    resolver: yupResolver(schema),
    mode: "onChange",
  });

  const Submit = async (inputData: FormValues) => {
    if (!isLoaded) {
      return;
    }

    await setActive({ session: null });

    try {
      const completeSignIn = await signIn.create({
        identifier: inputData.email,
        password: inputData.password,
      });

      if (completeSignIn.status === "complete") {
        await setActive({ session: completeSignIn.createdSessionId });
        navigate("/dashboard");
      }
    } catch (err) {
      const error = err as ErrorWithDetails;
      if (
        error.errors.some((error) => error.code === "form_identifier_not_found")
      ) {
        setSignInError("email", {
          type: "manual",
          message: "We can't find an account with that email address.",
        });
      }

      if (
        error.errors.some((error) => error.code === "form_password_incorrect")
      ) {
        setSignInError("password", {
          type: "manual",
          message: "Incorrect password.",
        });
      }

      await setActive({ session: null });
      console.error(JSON.stringify(err, null, 2));
    }
  };

  useEffect(() => {
    setDisabled(!isValid);
  }, [isValid]);

  return (
    <div className={styles.pageContainer}>
      <div className={styles.navbar}>
        <img src={logo} alt="Miracle Flights Logo" className={styles.logo} />
      </div>
      <div className={styles.contentContainer}>
        {/* Left side: login container */}
        <div className={styles.loginContainer}>
          <div className={styles.loginBlock}>
            <div className={styles.loginBlockHeader}>Login to your Account</div>
            <form
              onSubmit={handleSubmit(Submit)}
              className={styles.loginBlockContent}
            >
              <div className={styles.loginInputContainerUpper}>
                <Input
                  name="email"
                  register={register}
                  error={errors[FormValueNames.email]?.message}
                  type="text"
                  placeholder="Email"
                />
              </div>
              <div className={styles.loginInputContainerLower}>
                <Input
                  name="password"
                  register={register}
                  error={errors[FormValueNames.password]?.message}
                  type={passwordVisible ? "text" : "password"}
                  placeholder="Password"
                />
                <div
                  className={styles.passwordToggle}
                  onClick={() => setPasswordVisible(!passwordVisible)}
                >
                  <div className={styles.Icon}>
                    <Icon glyph={passwordVisible ? "eye-slash" : "eye"} />
                  </div>
                </div>
              </div>
              <div className={styles.forgotPassword}>
                <div
                  onClick={() => navigate("/forgot-password")}
                  className={styles.loginBlockLink}
                >
                  Forgot password?
                </div>
              </div>
              <Button
                variant={ButtonVariant.Login}
                type="submit"
                text="Login"
                disabled={disabled}
              />
            </form>
          </div>
        </div>

        <div className={styles.rightSideContent}>
          <div className={styles.rightSideHeader}>New Here?</div>
          <div className={styles.rightSideText}>
            Create your account with us today!
          </div>
          <Button
            text="Sign Up"
            type="button"
            variant={ButtonVariant.Signup}
            onClick={() => navigate("/sign-up")}
          />
        </div>
      </div>
    </div>
  );
};

export default LoginPage;
