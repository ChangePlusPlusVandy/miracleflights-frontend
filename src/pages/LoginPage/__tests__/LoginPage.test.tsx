import LoginPage from "../LoginPage";
import { render } from "@testing-library/react";

// mock the router provider
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

// mock clerk
//  const { signIn, setActive, isLoaded } = useSignIn();
jest.mock("@clerk/clerk-react", () => ({
  useSignIn: jest.fn(() => ({
    signIn: jest.fn(),
    setActive: jest.fn(),
    isLoaded: true,
  })),
}));

/**************************LOGIN PAGE TESTING SUITE*******************************/
describe("Login Page Tests", () => {
  /**************************SNAPSHOT TEST*******************************/
  test("Renders the Login Page snapshot correctly", () => {
    const { container } = render(<LoginPage />);
    expect(container).toMatchSnapshot();
  });
});
