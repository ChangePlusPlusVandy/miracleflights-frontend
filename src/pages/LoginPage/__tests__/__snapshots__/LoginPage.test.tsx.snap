// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Tests Renders the Login Page snapshot correctly 1`] = `
<div>
  <div
    class="pageContainer"
  >
    <div
      class="navbar"
    >
      <img
        alt="Miracle Flights Logo"
        class="logo"
        src="test-file-stub"
      />
    </div>
    <div
      class="contentContainer"
    >
      <div
        class="loginContainer"
      >
        <div
          class="loginBlock"
        >
          <div
            class="loginBlockHeader"
          >
            Login to your Account
          </div>
          <form
            class="loginBlockContent"
          >
            <div
              class="loginInputContainerUpper"
            >
              <div
                class="form-field"
              >
                <div
                  class="label-and-error"
                />
                <input
                  name="email"
                  placeholder="Email"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="loginInputContainerLower"
            >
              <div
                class="form-field"
              >
                <div
                  class="label-and-error"
                />
                <input
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="passwordToggle"
              >
                <div
                  class="Icon"
                >
                  <i
                    class="Icon fas fa-eye Text-fontSize--"
                  />
                </div>
              </div>
            </div>
            <div
              class="forgotPassword"
            >
              <div
                class="loginBlockLink"
              >
                Forgot password?
              </div>
            </div>
            <button
              class="button Login Blue"
              disabled=""
              type="submit"
            >
              <h4>
                Login
              </h4>
            </button>
          </form>
        </div>
      </div>
      <div
        class="rightSideContent"
      >
        <div
          class="rightSideHeader"
        >
          New Here?
        </div>
        <div
          class="rightSideText"
        >
          Create your account with us today!
        </div>
        <button
          class="button Signup Blue"
          type="button"
        >
          <h4>
            Sign Up
          </h4>
        </button>
      </div>
    </div>
  </div>
</div>
`;
